pkg/sql/catalog/descs/kv_descriptors.go

```diff
@@ -207,32 +207,50 @@ func (kd *kvDescriptors) getByName(
 	if !found || err != nil {
 		return found, nil, err
 	}
+	return kd.getDescriptor(ctx, txn, descID, mutable, name)
+}
+
+func (kd *kvDescriptors) getDescriptor(
+	ctx context.Context, txn *kv.Txn, id descpb.ID, mutable bool, maybeLookedUpName string,
+) (found bool, _ catalog.Descriptor, _ error) {
+	withNameLookup := maybeLookedUpName != ""
+	if id == keys.SystemDatabaseID {
+		b := dbdesc.NewBuilder(systemschema.SystemDB.DatabaseDesc())
+		// Special handling for the system database descriptor.
+		// The system database descriptor should never actually be mutated, which is
+		// why we return the same hard-coded descriptor every time. It's assumed
+		// that callers of this method will check the privileges on the descriptor
+		// (like any other database) and return an error.
+		if mutable {
+			return true, b.BuildExistingMutable(), nil
+		}
+		return true, b.BuildImmutable(), nil
+	}
 	// Always pick up a mutable copy so it can be cached.
-	desc, err = catalogkv.GetMutableDescriptorByID(ctx, txn, kd.codec, descID)
+	desc, err := catalogkv.MustGetMutableDescriptorByID(ctx, txn, kd.codec, id)
 	if err != nil {
+		if withNameLookup && errors.Is(err, catalog.ErrDescriptorNotFound) {
+			// Having done the namespace lookup, the descriptor must exist.
+			err = errors.WithAssertionFailure(err)
+		}
 		return false, nil, err
-	} else if desc == nil && descID <= keys.MaxReservedDescID {
-		// This can happen during startup because we're not actually looking up the
-		// system descriptor IDs in KV.
-		return false, nil, errors.Wrapf(catalog.ErrDescriptorNotFound, "descriptor %d not found", descID)
```

Comment:
`catalogkv` makes nice, uniformly-defined descriptor errors these days, so there's no point in making our own here.